### PR TITLE
Updating documentation with current doctest usage code and discussion for master branch

### PR DESCRIPTION
**Pull request by PaulWay**
This is the master branch version of #845

### CODE REVIEW

File: docs/customtut_parsers.rst [hunk 1]
```diff
@@ -457,125 +457,202 @@ The following shows our completed parser including documentation.
...
-If no errors are displayed then ``doctest`` was successful. To run
-``pytest`` on just the ``ssh`` parser execute the following command::
+    def test_sshd_config():
+        sshd_config = SshDConfig(context_wrap(SSHD_CONFIG_INPUT))
+        assert sshd_config is not None
+        assert 'Port' in sshd_config
+        assert 'PORT' in sshd_config
+        assert sshd_config['port'] == ['22', '22']
+        assert 'ListenAddress' in sshd_config
+        assert sshd_config['ListenAddress'] == ['10.110.0.1', '10.110.0.1']
```
Comment (psachin) [inline docs/customtut_parsers.rst]: This will fail as both items in list are same

File: docs/customtut_parsers.rst [hunk 2]
```diff
@@ -457,125 +457,202 @@ The following shows our completed parser including documentation.
...
+.. code-block:: python
+    :linenos:
+    :lineno-start: 51
+
+    def test_sshd_corner_cases():
+        """
+        Here we test any corner cases for behaviour we expect to deal with
+        in the parser but doesn't make a good example.
+        """
+        config = ssh.SshDConfig(context_wrap(SSHD_DOCS_EXAMPLE))
```
Comment (psachin) [inline docs/customtut_parsers.rst]: Works well with `SSHD_CONFIG_INPUT`, if we keep `SSHD_DOCS_EXAMPLE`, it may confuse users

File: docs/customtut_parsers.rst [hunk 3]
```diff
@@ -457,125 +457,202 @@ The following shows our completed parser including documentation.
...
+        assert len(ports) == 2
+        assert ports[0].value == '22'
+
+To test the documentation, we can then use ``doctest``:
+
+.. code-block:: python
+    :linenos:
+    :lineno-start: 35
+
+    def test_sshd_documentation():
```
Comment (psachin) [inline docs/customtut_parsers.rst]: 👍 for adding `test_sshd_documentation`

File: docs/customtut_parsers.rst [hunk 4]
```diff
@@ -457,125 +457,202 @@ The following shows our completed parser including documentation.
...
+        """
+        Here we test the examples in the documentation automatically using
+        doctest.  We set up an environment which is similar to what a
+        rule writer might see - a 'shared' object that stores parser
+        objects by class.  This saves doing this setup in the example code.
+        """
+        env = {
+            'SshDConfig': ssh.SshDConfig,
+            'shared': {
```
Comment (bfahr) [inline docs/customtut_parsers.rst]: 3.x no longer uses the `shared` mechanism so we should change this to reflect how it should be done in 3.x.

File: docs/docs_guidelines.rst [hunk 5]
```diff
@@ -47,64 +47,52 @@ The following code provides an example of how docstring comments should be used
...
+
+    The data is exposed via the ``obj.lines`` attribute which is a list containing
+    each line in the output.  The data may also be filtered using the
+    ``obj.get("filter string")`` method.  This method will return a list of lines
+    containing only "filter string".  The ``in`` operator may also be used to test
+    whether a particular string is in the ``lspci`` output.  Other methods/operators
+    are also supported, see the :py:class:`insights.core.LogFileOutput` class for more information.
+
+    Examples:
+        >>> pci_info = shared[LsPci]
```
Comment (bfahr) [inline docs/docs_guidelines.rst]: Same comment regarding `shared`.

File: docs/customtut_parsers.rst [hunk 6]
```diff
@@ -457,125 +457,202 @@ The following shows our completed parser including documentation.
...
+        Port 22
+        ListenAddress 10.110.1.1
+        #ListenAddress ::
+
+        # The default requires explicit activation of protocol 1
+        #Protocol 2
+        Protocol 1
+
+    Examples:
+        >>> sshd_config = shared[SshDConfig]
```
Comment (bfahr) [inline docs/customtut_parsers.rst]: See comment below regarding use of `shared` in 3.x.

File: docs/customtut_parsers.rst (hunk 1 above)
Comment (PaulWay) [reply docs/customtut_parsers.rst]: Got it, thanks!  Well spotted.

File: docs/customtut_parsers.rst (hunk 2 above)
Comment (PaulWay) [reply docs/customtut_parsers.rst]: `SSHD_DOCS_EXAMPLE` should actually be a separate definition, so I've included that.  Thanks!

File: docs/customtut_parsers.rst (hunk 3 above)
Comment (PaulWay) [reply docs/customtut_parsers.rst]: Thanks!

**Comment by PaulWay** [comment]
Hi @bfahr - Thanks for picking that up, I'd forgotten that point of course.  I've updated it now - let me know what you think.
